BED_MESH_TILT implementation

GarthSnyder opened the pull request:
This PR is intended to supersede PRs #2107 by @antst and #2018 by @jperrin72 (at jperrin72's request). All credit for this goes to @antst (Anton Starikov) - I just ported it to the current code base and added a few finishing touches.

This command probes three points on the bed to determine its current attitude and then adjusts
the currently-active bed leveling mesh to conform it to the actual state of the bed.

Pursuant to comments from @Arksine on PR #2107, meshes that have been tilted cannot be saved. Or rather, they can be saved, but it is the original, untilted probe values that go into the config file. I think this compromise will help to ensure that saved configurations include the highest quality data while preventing people from backing themselves into a corner (e.g., by making a complex map, testing it under a slicer profile that does a BED_MESH_TILT, and then discovering that the mesh can no longer be saved).

This patch fixes what I believe is likely an issue in the original master code which applies the bed_mesh settings from printer.cfg to every saved bed mesh. The params now come from the active mesh and are ordered canonically.

Includes docs.

This version continues to support `relative_reference_index`, but I believe it's now doing so correctly. 🤞

I have been testing this code for a while on my Sidewinder X1, and it has been working fine for me. Unfortunately, there are some delta-specific code paths that I'm unable to test. 

Open issues:

* Need to test on delta.

* Limited to 3 probe points, so probing error is potentially on the order of three times the error of one probe (which on my printer seems to be about 0.01). That seems like enough to affect mesh quality, although just barely.

* Is there any chance numpy could be added as a dependency? It's easily installed without compilation on all the platforms that have a Klipper install- script with the possible exception of beaglebone, although that's Linux too. There seems to be quite a bit of linear algebra in Klipper, and this feature is just one case in point. The [Kabsch algorithm](https://en.wikipedia.org/wiki/Kabsch_algorithm) would allow the current 3-point algorithm to be generalized to N points, and with numpy this is about [10 lines of code](https://math.stackexchange.com/questions/99299/best-fitting-plane-given-a-set-of-points) - shorter, in fact, than the current code.

* This PR is a single commit, which seems to be the standard for Klipper. But I have more detailed chains available if that's preferred.

jperrin72 commented:
>This PR is intended to supersede PRs #2107 by @antst and #2018 by @jperrin72 (at jperrin72's request). All credit for this goes to @antst (Anton Starikov) - I just ported it to the current code base and added a few finishing touches. ...

@GarthSnyder thanks for moving this forward 😀

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

KevinOConnor commented:
> Is there any chance numpy could be added as a dependency? It's easily installed without compilation on all the platforms that have a Klipper install- script with the possible exception of beaglebone, although that's Linux too. There seems to be quite a bit of linear algebra in Klipper, and this feature is just one case in point. The Kabsch algorithm would allow the current 3-point algorithm to be generalized to N points, and with numpy this is about 10 lines of code - shorter, in fact, than the current code.

Short answer: I haven't looked closely at the code, but you should be able to use `mathutil.coordinate_descent()` to solve for the tilt in only a few lines of code.  See how klippy/extras/bed_tilt.py does this.  Using a solver may not be the most mathematically elegant or cpu efficient, but it'll come up with a precise solution in a short time.

Long answer: Installing numpy via `pip install numpy` takes a very long time on a Raspberry Pi (last time I checked, I killed the install after something like 10 minutes).  The install is very fast via `sudo apt-get install`, but that installs the system version - which may be different on different machines.  Thus, using numpy is something I'd prefer to avoid unless it offers something essential.

-Kevin

Arksine commented:
I added a few comments above.  Thinking further about the z offset, could we just take an average of the tilt points, then offset them using that value?  It seems that the offset from the mesh isn't really important, what we are interested in is a change in the delta among the tilt points.

GarthSnyder commented on klippy/extras/bed_mesh.py:
```diff
@@ -215,13 +218,23 @@ def __init__(self, config, bedmesh):
         self.bedmesh = bedmesh
         self.probed_matrix = None
+        self.probed_matrix_backup = None
+        self.build_map = False
```
Fixed - I'll remove the other reference in my clone as well.

GarthSnyder commented on klippy/extras/bed_mesh.py:
```diff
@@ -484,6 +525,71 @@ def print_probed_positions(self, print_func):
+        z_in_mesh = self.bedmesh.z_mesh.calc_z(*pos[0:2]) + z_offset
+        delta_z = pos[2] - z_in_mesh;
+        return [pos[0], pos[1], delta_z]
+
```
Yes, that looks right, thanks. Fixed.

GarthSnyder commented on klippy/extras/bed_mesh.py:
```diff
@@ -484,6 +525,71 @@ def print_probed_positions(self, print_func):
+
+        if self.relative_reference_index is not None:
+            rri = self.relative_reference_index
+            z_offset = self.bedmesh.z_mesh.calc_z(*self.points[rri][0:2])
```
> Arksine: Thinking further about the z offset, could we just take an average of the tilt points, then offset them using that value? It seems that the offset from the mesh isn't really important, what we are interested in is a change in the delta among the tilt points.

Thanks for the thoughtful comments. 

I was a bit thrown off in my thinking by the `relative_reference_index` issue, as it seemed to suggest that the bed mesh system had some role in establishing the zero level for the Z axis as a whole. But reading more closely through the code, that appears to be an entirely separate mechanism and is controlled by the axis settings or the probe offset.

I believe you're right about offsets being irrelevant for meshes, at least as far as I can determine with my partial understanding of the code. Either the user will specify a fade target, which will act as a global offset, or the code will calculate an average Z value and use that as the offset when the mesh is activated. Offsets applied to incoming meshes are therefore irrelevant except to the extent that they might alter the value the user needs to specify as a fade target, if that feature is in use.

A corollary is that `relative_reference_index` doesn't really seem to have a functional role anywhere in the current codebase. In `probe_finalize` it specifies an alternative offset applied to incoming probe values (in lieu of the probe's own Z offset), but since the value is constant it has no real effect, as you outlined (and neither would the probe's Z offset). `tilt_probe_finalize` is just matching the general flow of `probe_finalize`, but the adjustment is equally unnecessary in both places, is it not?

I'm wondering if, in addition to normalizing probe matrices to an average vale of zero, it might be possible to just deprecate `relative_reference_index` and the code that uses it.

GarthSnyder commented:
> Arksine: Thinking further about the z offset, could we just take an average of the tilt points, then offset them using that value? It seems that the offset from the mesh isn't really important, what we are interested in is a change in the delta among the tilt points.

Thanks for the thoughtful comments. 

I was a bit thrown off in my thinking by the `relative_reference_index` issue, as it seemed to suggest that the bed mesh system had some role in establishing the zero level for the Z axis as a whole. But reading more closely through the code, that appears to be an entirely separate mechanism and is controlled by the axis settings or the probe offset.

I believe you're right about offsets being irrelevant for meshes, at least as far as I can determine with my partial understanding of the code. Either the user will specify a fade target, which will act as a global offset, or the code will calculate an average Z value and use that as the offset when the mesh is activated. Offsets applied to incoming meshes are therefore irrelevant except to the extent that they might alter the value the user needs to specify as a fade target, if that feature is in use.

A corollary is that `relative_reference_index` doesn't really seem to have a functional role anywhere in the current codebase. In `probe_finalize` it specifies an alternative offset applied to incoming probe values (in lieu of the probe's own Z offset), but since the value is constant it has no real effect, as you outlined (and neither would the probe's Z offset). `tilt_probe_finalize` is just matching the general flow of `probe_finalize`, but the adjustment is equally unnecessary in both places, is it not?

I'm wondering if, in addition to normalizing probe matrices to an average vale of zero, it might be possible to just deprecate `relative_reference_index` and the code that uses it.

GarthSnyder commented:
> KevinOConnor: `pip install numpy` takes a very long time on a Raspberry Pi...Thus, using numpy is something I'd prefer to avoid unless it offers something essential.

Ah yes, I see, that makes perfect sense. No, not essential. I figured you might have been down this road already... 🙂

It looks like there's a piwheel for numpy on Python3, but I don't see anything for P2.

I'll take a look at the coordinate descent implementation - that looks like a nice resource. It might simplify things even for the three-point case. Let me try it out before we merge anything.

Arksine commented:
Ah, I'll provide a deeper explanation of what relative reference index is.  To begin, it is only useful if you home the Z axis with an endstop.  When you calibrate the endstop, it will be against a specific x,y coordinate, and that coordinate will be Z=0.

What the relative reference index does is allow you to specify the index of one of the generated points that will become the "zero" reference for the mesh, ie the location that your calibrated your endstop against.  Essentially bed_mesh substitutes the probed valued value at a specified index for the probe's z_offset.  The index for each generated point can be found in the log, or by running `BED_MESH_OUTPUT PGP=1`.

GarthSnyder commented:
> Arksine: What the relative reference index does is allow you to specify the index of one of the generated points that will become the "zero" reference for the mesh, ie the location that your calibrated your endstop against. Essentially bed_mesh substitutes the probed valued value at a specified index for the probe's z_offset.

I understand that much, I think. But what is the value of what you've described given that mesh offsets are essentially arbitrary? Is it so you can look at the mesh in a visualizer and determine how it sits with respect to your chosen zero reference? Is it done with the expectation that it will let endstop-only folks use `fade_target: 0`?

Arksine commented:
Its primary purpose is to help accommodate probe temperature drift.  Some probes are more susceptible than others, and this is particularly true for enclosed printers like the Voron (the RRI patch was was added by a Voron member).  The probe is accurate enough in the short term, but at different temperatures the z_offset can differ significantly.  Rather than calibrate the z_offset for each temperature and reconfigure, the RRI allows the user to take the z-offset out of the equation.  You know where Z=0 should be, so whatever the probed value at that location is becomes the z-offset.

GarthSnyder commented:
I converted to the iterative solver and added a SAMPLES=<count> argument to BED_MESH_TILT. Coordinate descent works great, and I have cross-validated it against the old vector code, which is now archived.

Choosing a set of maximally-distributed points to probe turns out to be an unsolved problem isomorphic to circle-packing. This code uses static, best-known solutions from packomania.com scaled to fit the probeable area of the bed. A side benefit is that the probing paths are precomputed for minimal total length. There's a limit of 16 points, mostly to discourage people from thinking they ought to be using large Ns. My suspicion is that 3-8 points probably covers all reasonable scenarios.

As far as external code, I added a precision argument to `coordinate_descent()` since X and Y correction factors are typically on the order of 0.001. I also made Zmesh set the mesh_offset to 0 on build_mesh. And I added {min|max}_{x|y} to BedMeshCalibrate.mesh_params.

@Arksine, thanks again for your patience and help. Through experimentation, I've determined that the average value of the mesh _does_ in fact seem to (functionally) affect the global zero level. If you set a mesh with an average value of 100mm, the head will float around at 100m when it thinks it's at zero. The average value is subtracted from the internal representation of the mesh, but as you point out, it just goes into mesh_offset and is then added back as a constant. That mesh_offset part also becomes the fade_target, so it's effectively permanent even outside the compensation range. The variation in the mesh is faded out, but the average value persists as a permanent offset.

Given that, your explanation of relative_reference_index makes perfect sense. But now I'm reconfused about the original suggestion to center the adjustment mesh on zero. Isn't a nonzero average needed to correct for a bed that has changed its Z position?

Arksine commented:
The mesh itself is indeed sensitive to its height.  The Z-values in the mesh must be offset so so the mesh is properly set relative to Z=0.  By default the z-offset is used to do this, but when the relative_reference_index is set it uses the proved value at that coordinate the offset the mesh instead.

My proposal for using the average to offset the points was for the tilt points only.  My thinking was that we were only really interested in the normal, so the height of the tilt points shouldn't matter.   After taking a deeper look at the previous math that likely wouldn't have worked out.  Regardless it wouldn't apply to your new method.

I'll take a deeper look at what your new method of applying tilt as soon as I get a chance.  I did look at your changes to the ZMesh class. I think its correct to reset the `mesh_offset` to 0 now since we are calling build_mesh() more than once on a mesh.  The changes to offset_mesh() are unnecessary is it isn't valid to call offset_mesh() more than once after the mesh has been built.  The mesh_offset isn't a value we want to accumulate.

GarthSnyder commented:
Thanks, @Arksine. I pushed a couple more small fixes to address your most recent comments.

It sounds like there should in fact be some kind of adjustment based on the relative reference index; that's not in there yet.

jperrin72 commented:
Thanks @GarthSnyder @Arksine for this great work. Any chance to commit this soon ? 😃

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

GarthSnyder commented:
> Thanks @GarthSnyder @Arksine for this great work. Any chance to commit this soon ? 😃

I think @Arksine was going to take a more detailed look at the code first.

I believe everything's working correctly - it's pretty much ready to go modulo possible changes for relative reference index. If you want to pull the branch and run it, @jperrin72, it should be safe to do so.

Arksine commented:
Hello everyone, sorry for the delay.   I had a chance to take a closer look at this and the bed_mesh portion looks good.   Regarding the relative_reference_index, I think there are two viable options:

1)  Disregard the RRI in BED_MESH_TILT and use the probe's z_offset.  We could document that BED_MESH_TILT requires a properly calibrated probe z_offset even if the RRI is configured.

2)  If the RRI is configured we can append the coordinate at the RRI to the tilt points, then use the value probed at that location to offset the actual tilt points.  This would essentially mimic the way the RRI is used when the mesh is originally generated.  

I'm open to either method, the first is easier (nothing to implement or test) and the second is more robust.

jperrin72 commented:
>Hello everyone, sorry for the delay.   I had a chance to take a closer look at this and the bed_mesh portion looks good.   Regarding the relative_reference_index, I think there are two viable options: ...

@Arksine Thanks - I believe this is going to be a major feature for a lot of printer’s owners! Especially those with a poorly designed Y moving bed. 😊

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

GarthSnyder commented:
Thanks once again, @Arksine. I'm all for robustness and correctness - I'm just hampered a bit by my incomplete understanding of how all the pieces fit together.

Here's an RRI fix. It's just a couple lines of code with a long explanation. 🙂 It is essentially your #2 above; however, the BED_MESH_TILT probe points and the BED_MESH_CALIBRATE probe points are unrelated, so the set of BED_MESH_TILT probe points will likely not include the actual RRI point.

So instead, the adjustment is done in correction space. The adjustment is done in terms of absolute coordinates, so it's easy enough to calculate how the RRI point would be tilted. That scalar is then subtracted from all adjustments to leave the RRI point with a net tilt adjustment of zero.

I modified the existing `_sample_direct()` to make a copy of the probe matrix. If it does not, then both the BedMesh's and BedMeshCalibrate's copies of the original probe matrix are offset when `offset_mesh()` is called, corrupting the probe history. `_sample_bicubic()` and `_sample_lagrange()` are already creating new matrices.

I changed the default number of BED_MESH_TILT probe points to 4 instead of 3. That gives full-area coverage of rectangular beds.

Barring discovery of more issues, I think this PR is ready to go. Should I recompress it into a single commit?

Arksine commented:
Your method for adjusting the tilted mesh relative to the RRI looks good.  I also agree with the change to make a deep copy of the matrix when sampling direct.   As far as the number default points go, intuitively I would have thought 3 is good, however I'll leave that up to you since you are in a position to test and determine what the best defaults are.  From my perspective its good to merge.

Perhaps @KevinOConnor can chime in to let you know if you should rebase/fixup the commits before merging?

GarthSnyder commented:
> As far as the number default points go, intuitively I would have thought 3 is good, however I'll leave that up to you since you are in a position to test and determine what the best defaults are.

On my printer, there's quite a jump between 3 and 4, perhaps more than one might expect. (Here, the "right" answer at each point is the average of ten 16-point correction runs.) It does make a bit of intuitive sense given that 4 is the lowest number of samples that allows for some redundancy. The shape of the SSE curve is very similar.

![image](https://user-images.githubusercontent.com/443181/78593339-5254b080-77fb-11ea-9423-6f6ef9544896.png)

jperrin72 commented:
> > As far as the number default points go, intuitively I would have thought 3 is good, however I'll leave that up to you since you are in a position to test and determine what the best defaults are.
> 
> On my printer, there's quite a jump between 3 and 4, perhaps more than one might expect. (Here, the "right" answer at each point is the average of ten 16-point correction runs.) It does make a bit of intuitive sense given that 4 is the lowest number of samples that allows for some redundancy. The shape of the SSE curve is very similar.
> 
> ![image](https://user-images.githubusercontent.com/443181/78593339-5254b080-77fb-11ea-9423-6f6ef9544896.png)

indeed - too bad I'm quarantined away from my 3d printer now - can't wait to be back home to test it!

KevinOConnor commented:
Thanks.  I can squash all of these commits into one commit during merge.  If you'd prefer there be multiple separate commits then you would need to rebase the commits and update this PR.  If you choose the latter, make sure every commit message follows the guidelines described at: https://github.com/KevinOConnor/klipper/blob/master/docs/CONTRIBUTING.md

-Kevin

GarthSnyder commented:
This really isn't a lot of code - I'll keep a separate copy of the full history.

Collapsed to one commit and groomed for compliance, should be ready to go.

KevinOConnor commented:
Thanks.  One thing I'm unsure of is the addition of tilt_points.py.  Can you provide further details on what this does and why it is needed?  I'm guessing that this is to choose points for probing, but I wonder if this is a lot of complexity.  Wouldn't we want to probe at points that were used during the mesh probing (so as to reduce skew caused by mesh interpolation)?  If we reuse probe mesh points, then I would think a heuristic that chose corner points (for rectangles; XY extremes for circles) would provide a better solution as the tilt would be easier to measure with those extreme points.

-Kevin

GarthSnyder commented:
> KevinOConnor wrote: One thing I'm unsure of is the addition of tilt_points.py. Can you provide further details on what this does and why it is needed?

This is just a data table. The point sets in this file answer the question "Given a circular or rectangular boundary, what is the most widely distributed set of N points within that boundary?" Or more specifically, "What set of N points maximizes the minimum distance between any two points?" These layouts maximize the information obtained from each sample; they can't be improved upon.

The reason these are static layouts (as opposed to some algorithm that picks points) is that no algorithm to answer this question is known. There is not even a brute-force algorithm; the problem is isomorphic to [circle-packing](https://en.wikipedia.org/wiki/Circle_packing_in_a_square). All that exists is a catalog of the best solutions so far discovered.

> I would think a heuristic that chose corner points (for rectangles; XY extremes for circles) would provide a better solution as the tilt would be easier to measure with those extreme points.

You would think so, wouldn't you! But unfortunately, no, it's not the case. For N = 4, picking the four corners is optimal, but for N = 3, the optimal solution includes only one corner point. If you pick three corners, you're ignoring fully half of a rectangular bed. At large N, confining probe points to the boundary just squeezes them into close proximity, rendering the additional points largely noncontributory.

I'm sure it would be possible to come up with _some_ ad-hoc way of picking points that runs in reasonable time, especially if the possible point options were discretized. But in addition to giving known-to-be-inferior results, this option actually seems quite a bit more complex to me. Isn't static data with no code simpler and more bulletproof? The bottom line is that this is a harder problem than it might initially seem.

> Wouldn't we want to probe at points that were used during the mesh probing (so as to reduce skew caused by mesh interpolation)?

I can imagine that being an issue if the bed has local (between-probe-point) height variations that are large in comparison with the intrinsic accuracy of the probe. But I wonder how common that scenario is in real life; not very, I suspect. If the bed is locally smooth and every probe point has equal accuracy, an interpolated value shouldn't really be any less accurate than the grid points surrounding it. In fact, it should be slightly more accurate since at least some portion of the error at grid points is likely to be uncorrelated.

If you think this is likely to be an issue, I would imagine the best option would be to keep the current probe schedule and snap individual probe points to their closest grid points. But in some common cases, this is going to degrade accuracy considerably. If the grid is 3x3 and BED_MESH_TILT needs 3 samples, it's going to degrade to the case where half the bed is ignored.

My main concern with the static data table approach is the memory cost of storing the probing schedule. There are roughly 550 floating point values in 300 lists, and raw Python doesn't seem to be terribly efficient at storing this kind of data structure. The current version just has a static table, but I initially coded it so the relevant values would be read out of the data file as needed. In the end, since the value of additional probe points declines pretty rapidly, I just limited N to 16 and left the static table in. If you prefer, we could either put back the read-in-passing code or lower max(N) to something like 10.

KevinOConnor commented:
> You would think so, wouldn't you! But unfortunately, no, it's not the case. For N = 4, picking the four corners is optimal, but for N = 3, the optimal solution includes only one corner point. If you pick three corners, you're ignoring fully half of a rectangular bed. 

I feel I'm missing something significant.  I thought the purpose of this PR was to adapt a previous measured bed mesh to a new mesh where the bed is tilted at a different angle.  The goal being that one could save time if the bed is unchanged but becomes tilted - instead of probing the whole mesh again, one could probe a handful of points to obtain the tilt and then reuse the existing mesh.

So, using an example, lets consider a 9x9 mesh and of those 81 points, point A at 10,10 measured Z=0.0 and point B at 190,10 measured Z=0.0 .  If the code then later measured point A at Z=0.0 and point B at Z=0.9, then the mesh can be tilted with an X slope of .005 (== 0.9 / 180).

I don't see how picking the corner points reduces the effectiveness of measuring the tilt.  It would seem to me it is the optimal method of measuring the tilt.

-Kevin

jperrin72 commented:
@KevinOConnor Hi Kevin, Do you think we could release a first version with @GarthSnyder approach to defining corner points for broader testing and feedback from the users ?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

GarthSnyder commented:
I'm working on a response - sorry it's been a few days. I don't think there's much benefit to releasing potentially non-final code at this point. It shouldn't be much longer...

KevinOConnor commented:
Any further updates on this?

-Kevin

GarthSnyder commented:
> @KevinOConnor wrote: Any further updates on this?

I'm still actively working on this. I wanted to have an analytically solid argument one way or the other before responding, so I developed a sampling-based whole-bed error model which I've since migrated to a partially closed form model that uses SymPy. This is actually a very interesting problem.

It was clear early on that you were right about the (low) value of interior points; moving them toward the periphery always improves the whole-bed error score. But the wrinkle here is that this is true in the extreme: with a straightforward error model, nothing is as valuable as corner points. If you have the opportunity to sample 16 points, this model says, you should sample each of the corners 4 times each and ignore everything else.

What that model ignores is that there are really two sources of error in the tilt calculation. Not only are you sampling the bed, but you're also comparing your samples against an existing bed mesh, which itself contains some sampling error. If you sample a corner 4 times, you are averaging out some of the immediate sampling error, but you're also multiplying the effect of the error built into the original bed mesh. Perhaps counterintuitively, it seems to be advantageous to sample in places where the original mesh _wasn't_ sampled, because interpolation (of the mesh) has the effect of balancing out some of the sampling errors in the original mesh.

I think the optimal pattern is going to be something along the lines of "points distributed evenly around the perimeter of the bed, then inset by half the size of the current mesh's grid, then (for rectangular beds) skewed toward the corners, but not too close to each other." I'm just trying to iron out specifics and make sure it's not too dependent on any particular model assumptions.

The obvious question is whether it's really worthwhile to fuss over this in such detail. It does seem to be - the general pattern described above scores about 25% better than the current model or any of the obvious regular patterns. (The error metric being the sum of squared errors integrated across the bed.)

If you would like to get this out, even distribution around the bed periphery seems to be as good a stopgap as any (even if that layout misses the corners, oddly enough). I can put that in for now. It seems pretty clear that the eventual solution will be some kind of layout procedure, not a table.

KevinOConnor commented:
Thanks.

> What that model ignores is that there are really two sources of error in the tilt calculation.

FWIW, I suspect that type of error would be overwhelmed by "human errors".  That is, I fear additional complexity would cause more errors than would be gained in the "math".

My suggestion would be to have `BED_MESH_TILT` take no parameters (and no additional configuration).  Have the code select a sane set of points from the previously probed mesh, probe them again, calculate the tilt, and then apply the tilt to the mesh.  For example, select the corners on a square mesh, the extremes on a round mesh, or some other simple and easy to describe algorithm for point selection.  I'd guess somewhere between 2 and 4 points (your choice) would be sane for real-world usage.  This way it is easy to describe the behaviour and users/developers aren't asked to make/maintain a choice they don't need to make.

Cheers,
-Kevin

jperrin72 commented:
Can I also suggest that some kind of error threshold is implemented into the procedure so that the print job can be canceled when the tilt compare to the original mesh is too important. This is especially important for multi-extruder tool head that would cause inactive nozzle to bump into the layers if the bed leveling is really gone.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

GarthSnyder commented:
> KevinOConnor wrote: My suggestion would be to have `BED_MESH_TILT` take no parameters (and no additional configuration). Have the code select a sane set of points from the previously probed mesh, probe them again, calculate the tilt, and then apply the tilt to the mesh...I'd guess somewhere between 2 and 4 points (your choice) would be sane for real-world usage.

For this scenario, I think you're probably best off going back to @antst's original code in PR #2107. That implementation is actually pretty complete and does exactly what you want.

jperrin72 commented:
@KevinOConnor hi Kevin, unfortunate it is the third attempt to enhance bed mesh tilt implementation that fall short. How do you think we can move from there ? Thanks Jérôme 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

jperrin72 commented:
I propose to put together both PR into a new one - will start working on it in a couple of weeks

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

MironV commented:
What is the correct PR tracking this work? Sorry, I'm having a hard time finding the right one and this feature is pretty important.

andreasmwestberg commented:
Is there any update on this or somewhere it's tracked? I cannot find anything
